helper/schema/resource_diff.go

```diff
@@ -0,0 +1,446 @@
+			Map:    BasicMapReader(stateAttributes),
+		}
+	}
+	if d.config != nil {
+		readers["config"] = &ConfigFieldReader{
+			Schema: d.schema,
+			Config: d.config,
+		}
+	}
+	if d.diff != nil {
+		readers["diff"] = &DiffFieldReader{
+			Schema: d.schema,
+			Diff:   d.diff,
+			Source: &MultiLevelFieldReader{
+				Levels:  []string{"state", "config"},
+				Readers: readers,
+			},
+		}
+	}
+	readers["newDiff"] = &newValueReader{
+		MapFieldReader: &MapFieldReader{
+			Schema: d.schema,
+			Map:    BasicMapReader(d.newWriter.Map()),
+		},
+		computedKeys: d.newWriter.ComputedKeysMap(),
+	}
+	d.multiReader = &MultiLevelFieldReader{
+		Levels: []string{
+			"state",
+			"config",
+			"diff",
+			"newDiff",
+		},
+
+		Readers: readers,
+	}
+
+	d.updatedKeys = make(map[string]bool)
+
+	return d
+}
+
+// UpdatedKeys returns the keys that were updated by SetNew, SetNewComputed, or
+// SetDiff. These are the only keys that a diff should be re-calculated for.
+func (d *ResourceDiff) UpdatedKeys() []string {
+	var s []string
+	for k := range d.updatedKeys {
+		s = append(s, k)
+	}
+	return s
+}
+
+// Clear wipes the diff for a particular key. It is called by SetDiff to remove
+// any possibility of conflicts, but can be called on its own to just remove a
+// specific key from the diff completely.
+//
+// Note that this does not wipe an override. This function is only allowed on
+// computed keys.
+func (d *ResourceDiff) Clear(key string) error {
+	if !d.schema[key].Computed {
+		return fmt.Errorf("Clear is allowed on computed attributes only - %s is not one", key)
+	}
+
+	return d.clear(key)
+}
+
+func (d *ResourceDiff) clear(key string) error {
+	// Check the schema to make sure that this key exists first.
+	if _, ok := d.schema[key]; !ok {
+		return fmt.Errorf("%s is not a valid key", key)
+	}
+	for k := range d.diff.Attributes {
+		if strings.HasPrefix(k, key) {
+			delete(d.diff.Attributes, k)
+		}
+	}
+	return nil
+}
+
+// diffChange helps to implement resourceDiffer and derives its change values
+// from ResourceDiff's own change data, in addition to existing diff, config, and state.
+func (d *ResourceDiff) diffChange(key string) (interface{}, interface{}, bool, bool) {
+	old, new := d.getChange(key)
+
+	if !old.Exists {
+		old.Value = nil
+	}
+	if !new.Exists {
+		new.Value = nil
+	}
+
+	return old.Value, new.Value, !reflect.DeepEqual(old.Value, new.Value), new.Computed
+}
+
+// SetNew is used to set a new diff value for the mentioned key. The value must
+// be correct for the attribute's schema (mostly relevant for maps, lists, and
+// sets). The original value from the state is used as the old value.
+//
+// This function is only allowed on computed attributes.
+func (d *ResourceDiff) SetNew(key string, value interface{}) error {
+	if !d.schema[key].Computed {
+		return fmt.Errorf("SetNew only operates on computed keys - %s is not one", key)
+	}
+	return d.setDiff(key, d.get(strings.Split(key, "."), "state").Value, value, false)
```

Comment:
During some testing in which I was prototyping using these features in the Nomad provider, I noticed that if you call `SetNew` with a key that doesn't exist in the schema you can get a panic:

```
2017/05/31 15:36:04 [DEBUG] plugin: terraform: [signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0xace9b1]
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 
2017/05/31 15:36:04 [DEBUG] plugin: terraform: goroutine 72 [running]:
2017/05/31 15:36:04 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*ResourceDiff).SetNew(0xc4203f6940, 0x51fdf73, 0x2, 0x44cba80, 0xc4202a13f0, 0x0, 0x7c67d80)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/home/mart/Devel/terraform/src/github.com/hashicorp/terraform/helper/schema/resource_diff.go:259 +0x71
2017/05/31 15:36:04 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/nomad.resourceJobCustomizeDiff(0xc4203f6940, 0x50303a0, 0xc42059d480, 0xc4200c5300, 0xc4203f6940)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/home/mart/Devel/terraform/src/github.com/hashicorp/terraform/builtin/providers/nomad/resource_job.go:77 +0x156
2017/05/31 15:36:04 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.schemaMap.Diff(0xc42055aba0, 0xc4201e0f00, 0xc420535f50, 0x5399a80, 0x50303a0, 0xc42059d480, 0x1, 0x18, 0x28)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/home/mart/Devel/terraform/src/github.com/hashicorp/terraform/helper/schema/schema.go:413 +0xb9d
2017/05/31 15:36:04 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Resource).Diff(0xc420329420, 0xc4201e0f00, 0xc420535f50, 0x50303a0, 0xc42059d480, 0x1, 0x28, 0xc420418030)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/home/mart/Devel/terraform/src/github.com/hashicorp/terraform/helper/schema/resource.go:225 +0x187
2017/05/31 15:36:04 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Provider).Diff(0xc420329490, 0xc4201e08c0, 0xc4201e0f00, 0xc420535f50, 0x7f75821c1e10, 0x0, 0x0)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/home/mart/Devel/terraform/src/github.com/hashicorp/terraform/helper/schema/provider.go:255 +0x9b
2017/05/31 15:36:04 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Diff(0xc42015a980, 0xc42026b840, 0xc420429600, 0x0, 0x0)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/home/mart/Devel/terraform/src/github.com/hashicorp/terraform/plugin/resource_provider.go:499 +0x57
2017/05/31 15:36:04 [DEBUG] plugin: terraform: reflect.Value.call(0xc420362720, 0xc420434240, 0x13, 0x51ff043, 0x4, 0xc4205e7f20, 0x3, 0x3, 0x53a3bc8, 0xc4200286e8, ...)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/usr/lib/go-1.8/src/reflect/value.go:434 +0x91f
2017/05/31 15:36:04 [DEBUG] plugin: terraform: reflect.Value.Call(0xc420362720, 0xc420434240, 0x13, 0xc420028720, 0x3, 0x3, 0xc42002870c, 0x180001, 0x300000000)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/usr/lib/go-1.8/src/reflect/value.go:302 +0xa4
2017/05/31 15:36:04 [DEBUG] plugin: terraform: net/rpc.(*service).call(0xc4201c13c0, 0xc4201c1380, 0xc420428e40, 0xc420576600, 0xc42026b060, 0x43484c0, 0xc42026b840, 0x16, 0x4348500, 0xc420429600, ...)
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/usr/lib/go-1.8/src/net/rpc/server.go:387 +0x144
2017/05/31 15:36:04 [DEBUG] plugin: terraform: created by net/rpc.(*Server).ServeCodec
2017/05/31 15:36:04 [DEBUG] plugin: terraform: 	/usr/lib/go-1.8/src/net/rpc/server.go:481 +0x404
2017/05/31 15:36:04 [DEBUG] plugin: /home/mart/Devel/terraform/bin/terraform: plugin process exited
```

The panic went away once I realized my mistake in naming the attribute, but if possible it'd be nice for this to return an error rather than crash. :grinning: